File: On_Site_test_1.py
SIC_Program = list()
current_Address = 0
start_Address = int()
SYMTAB = dict() # string, int

Opcode = { #SIC模式下理論上opcode不會被動到 故直接存字串
    "Add": "18",
    "AND": "40",
    "COMP": "28",
    "DIV": "24",
    "J": "3C",
    "JEQ": "30",
    "JGT": "34",
    "JLT": "38",
    "JSUB": "48",
    "LDA": "00",
    "LDCH": "50",
    "LDL": "08",
    "LDX": "04",
    "MUL": "20",
    "OR": "44",
    "RD": "D8",
    "RSUB": "4C",
    "STA": "0C",
    "STCH": "54",
    "STL": "14",
    "STSW": "E8",
    "STX": "10",
    "SUB": "1C",
    "TD": "E0",
    "TIX": "2C",
    "WD": "DC"
}

class SIC_Line: #指令類別
    def __init__(self, line: str):
        print(line)
        self.address_Label, self.mnemonic_Opcode = line.removesuffix('\n').split('\t')[:2] #address_Label, mnemonic_Opcode, operand
        self.operands = line.removesuffix('\n').split('\t')[2] if len(line.removesuffix('\n').split('\t')) > 2 else ""
        self.object_Code = str()
        self.address = 0
        self.isDirectives = True; # 是否為組譯器指引(虛指令)

    def cal_Size(self): #計算大小同時生成虛指令的物件碼
        global start_Address
        self.length = 0
        match self.mnemonic_Opcode:
            case 'START':
                start_Address = int(self.operands, 16)
                self.address = start_Address
                self.object_Code = 'H' + self.address_Label.ljust(6, ' ') + f'{start_Address:x}'.rjust(6, '0')
                self.length = int(self.operands, 16)
            case 'END':
                self.object_Code = 'H' + f'{start_Address:x}'.rjust(6, '0')
                self.length = 0
            case 'BYTE':
                if self.operands.startswith('X'):
                    self.object_Code = self.operands[2:-1]
                if self.operands.startswith('C'):
                    self.object_Code = self.operands[2:-1].encode().hex()
                self.length = len(self.object_Code) // 2
            case 'WORD':
                self.object_Code = str(int(self.operands)).zfill(6)
                self.length = 3
            case 'RESB':
                self.length = int(self.operands)
            case 'RESW':
                self.length = int(self.operands) * 3
            case _:
                #實指令皆為3byte (format 1/2/4指令 只存在於XE)
                self.isDirectives = False
                return 3
        return self.length #不應該執行到這裡
    
    def pass2(self):
        if self.isDirectives:
            return
        addressing_Mode = 0      
        if self.operands != '' and self.operands.endswith(',X'):
            addressing_Mode = 0x8000
            self.operands = self.operands.removesuffix(',X')
        self.object_Code = Opcode[self.mnemonic_Opcode] + hex(SYMTAB[self.operands] + addressing_Mode).rjust(4, '0').removeprefix('0x')
    
    def pass1(self): #計算位置存入SYSTAB 處理虛指令
        global current_Address
        self.address = current_Address
        current_Address += self.cal_Size() #下一個指令的起始位置
        if self.address_Label != '':
            SYMTAB.update({self.address_Label: self.address})
        
if __name__ == "__main__":

    with open("input1.txt", "r") as f:
        SIC_Program = [SIC_Line(i) for i in f.readlines() if not i.startswith('.')] #將每一行的字串轉換成list並存入SIC_Program中,並直接移除註解部分
        
    #執行pass1計算loc以及SYMTAB
    for i in SIC_Program:
        i.pass1()
    SYMTAB.update({'': 0}) #空字串代表沒有label
    for i in SIC_Program:
        i.pass2()   
    
    with open("output1.txt", "w") as f:
        f.write(SIC_Program[0].object_Code + f'{(current_Address - start_Address):x}'.rjust(6, '0').upper() + '\n')
        
        line_Start_Address = -1; line_Object_Code = str()
        
        for i in range(1, len(SIC_Program)-1):
            print(SIC_Program[i].object_Code)
            if SIC_Program[i].object_Code == '':
                continue #如果為空則跳過
            if line_Start_Address == -1:
                line_Start_Address = SIC_Program[i].address #紀錄該行開頭的位置
                
            line_Object_Code += SIC_Program[i].object_Code
            if (len(line_Object_Code) + len(SIC_Program[i+1].object_Code)) > 60 or SIC_Program[i+1].object_Code == '' or SIC_Program[i+1].mnemonic_Opcode == 'END': #若長度達到極限(加上下一筆超過60個字元) 或是下一行並無物件碼/為END則輸出
                f.write('T' + hex(line_Start_Address).removeprefix('0x').rjust(6, '0').upper() + hex(len(line_Object_Code)//2).removeprefix('0x').rjust(2, '0').upper() + line_Object_Code.removeprefix('0x').upper() + '\n')
                line_Start_Address = -1
                line_Object_Code = str()
                
        f.write(SIC_Program[-1].object_Code.upper())

File: On_Site_test_1.cpp
#include <iostream>
#include<iomanip>
#include<fstream>
#include<string>
#include<sstream>
#include<vector>
#include<map>
using namespace std;
class SIC_Line;
int current_Address=0;
int start_Address;
ofstream output("output.txt", ios::out);
ofstream loc("loc.txt", ios::out);
vector<SIC_Line> SIC_Program;
map<string, int> SYMTAB;
map<string, string> Opcode = {
    {"ADD","18"},{"AND","40"},{"COMP","28"},{"DIV","24"},{"J","3C"},{"JEQ","30"},{"JGT","34"},{"JLT","38"},{"JSUB","48"},{"LDA","00"},{"LDCH","50"},{"LDL","08"},{"LDX","04"},{"MUL","20"},{"OR","44"},{"RD","D8"},{"RSUB","4C"},{"STA","0C"},{"STCH","54"},{"STL","14"},{"STSW","E8"},{"STX","10"},{"SUB","1C"},{"TD","E0"},{"TIX","2C"},{"WD","DC"}
};//SIC模式下理論上opcode不會被動到 故直接存字串
class SIC_Line
{
public:
    string address_Label = "";
    string mnemonic_Opcode = "";
    string operands = "";
    string object_Code="";
    int address=0;
    bool isDirectives = true;//是否為組譯器指引(虛指令)
    SIC_Line(string line)
    {
        stringstream ss;
        ss << line;
        getline(ss, address_Label, '\t');
        getline(ss, mnemonic_Opcode, '\t');
        getline(ss, operands, '\t');
    }
    void print()//偵錯用
    {
        cout<< setw(4) << setfill('0') << hex << address;
        cout << "\t" << address_Label << "\t" << mnemonic_Opcode << "\t" << operands << "\t" << object_Code << endl;
    }
    void pass1()//計算位置存入SYSTAB 處理虛指令
    {
        stringstream ss;
        address = current_Address;
        current_Address += cal_Size();//下一個指令的起始位置
        if (address_Label != "")
        {
            SYMTAB[address_Label]=address;
        }
        loc << uppercase << setw(4) << setfill('0') << hex << address<<"\t";
        loc << address_Label << "\t" << mnemonic_Opcode << "\t" << operands << endl;
    }
    int cal_Size()//計算大小同時生成虛指令的物件碼
    {
        //先判斷虛指令 start直接將起始位置視為size做加法
        if (mnemonic_Opcode == "START")
        {
            stringstream ss;
            start_Address= stoi(operands, 0, 16);
            address = start_Address;
            ss << "H"<< setw(6) << setfill(' ') << left<<address_Label<< setw(6) << setfill('0') << hex<<right << start_Address;
            getline(ss,object_Code);
            return stoi(operands, 0, 16);//將operands以16進制解讀 並回傳/載入到第二個參數位置中(未使用到故填入NULL(0))
        }
        if (mnemonic_Opcode == "END")
        {
            stringstream ss;
            ss << "E" << setw(6) << setfill('0') << hex << start_Address;
            ss >> object_Code;
            return 0;
        }
        if (mnemonic_Opcode == "BYTE")
        {
            if (operands[0] == 'X')
            {
                for (int i = 2; i < operands.length() - 1; i++)object_Code += operands[i];
            }
            if (operands[0] == 'C')
            {
                stringstream ss;
                for (int i = 2; i < operands.length() - 1; i++)
                {
                    ss << uppercase << setw(2) << hex << setfill('0') << (int)operands[i];
                }
                ss >> object_Code;
            }
            return object_Code.length() / 2;
        }
        if (mnemonic_Opcode == "WORD")
        {
            stringstream ss;
            ss << uppercase << hex<< setw(6) << setfill('0')  <<stoi(operands);
            ss >> object_Code;
            return 3;
        }
        if (mnemonic_Opcode == "RESB")
        {
            return stoi(operands);
        }
        if (mnemonic_Opcode == "RESW")
        {
            return stoi(operands) * 3;
        }
        //實指令皆為3byte (format 1/2/4指令 只存在於XE)
        isDirectives = false;
        return 3;
    }
    void pass2()
    {
      
        if(isDirectives==false)
        {
            int addressing_Mode = 0;
            if (operands!="" && operands[operands.length() - 2] == ',')//若倒數第二個為','(,X結尾 不會出現其他例外) 將索引定址模式設為1(0x8000) 
            {
                stringstream ss;
                ss << operands;
                getline(ss, operands, ',');//進行分割 去除",X"
                addressing_Mode = 0x8000;
            }
            stringstream ss;
            ss << hex << uppercase << setw(4) << setfill('0') << (SYMTAB[operands]+addressing_Mode);//先將address欄位輸出
            ss >> object_Code;
            object_Code = Opcode[mnemonic_Opcode]+ object_Code;//再拼上opcode
        }
        output <<uppercase<< setw(4) << setfill('0') << hex << address << "\t";//輸出 output.txt
        output << address_Label << "\t" << mnemonic_Opcode << "\t" << operands;
        if (!isDirectives || mnemonic_Opcode == "BYTE" || mnemonic_Opcode == "WORD")output << "\t" << object_Code;//只有實指令 BYTE WORD 需要輸出物件碼
        output << endl;
    }
};
int main()
{
    ifstream input_File("input.txt",ios::in);
    if (input_File.is_open())
    {
        string input_Line;
        while (getline(input_File, input_Line))
        {
            if(input_Line[0]!='.')SIC_Program.push_back(SIC_Line(input_Line));//非註解則輸入
        }
        input_File.close();
    }
    for (auto& L : SIC_Program)L.pass1();//執行pass1計算loc以及SYMTAB
    for (auto& L : SIC_Program)L.print();//偵錯用
    for (auto& L : SIC_Program)L.pass2();
    for (auto& L : SIC_Program)L.print();//偵錯用
    ofstream objectcode("objectcode.txt", ios::out);
    if (objectcode.is_open())
    {
        objectcode << uppercase << SIC_Program[0].object_Code << hex << setw(6) << setfill('0') << current_Address - start_Address << endl;//輸出H紀錄 分別為( H+程式名稱(6格寬)+起始位置 )+ 總程式長度 前者已於PASS1完成
        //T紀錄為 T+起始位置+長度(BYTE)+物件碼
        int line_Start_Address = -1;
        string line_Object_Code = "";
        for (int i = 1; i < SIC_Program.size() - 1; i++)
        {
            if (SIC_Program[i].object_Code == "")continue;//如果為空則跳過
            if (line_Start_Address == -1)line_Start_Address = SIC_Program[i].address;//若為開頭(尚無紀錄) 則紀錄該行開頭的位置
            line_Object_Code += SIC_Program[i].object_Code;//串接物件碼
            if (line_Object_Code.length() +SIC_Program[i+1].object_Code.length()>60 || SIC_Program[i + 1].object_Code == "" || SIC_Program[i + 1].mnemonic_Opcode == "END")//若長度達到極限(加上下一筆超過60個字元/30BYTE) 或是下一行並無物件碼/為END則輸出 
            {
                objectcode << "T" << uppercase << setw(6) << setfill('0') << hex << line_Start_Address << setw(2) << setfill('0') << hex << line_Object_Code.length() / 2 << line_Object_Code<<endl;
                //重置資料
                line_Start_Address = -1;
                line_Object_Code = "";
            }
        }
        objectcode << uppercase << SIC_Program[SIC_Program.size() - 1].object_Code;//輸出E紀錄
        objectcode.close();
        output.close();
        loc.close();
    }
    cin.get();
    return 0;
}
